### PR TITLE
Fix requires_grad of input must be true for activation checkpoint layer in pipeline train.

### PR DESCRIPTION
**Pull request by inkcherry**
mechanism FYI https://github.com/prigoyal/pytorch_memonger/blob/master/tutorial/Checkpointing_for_PyTorch_models.ipynb?short_path=e1e8ee7#L282C244-L282C257

For discrete input, could make activation for the embedding layer.
For image input,  no need to calculate and save grads for input.

Have verified on megatron-deepspeed train/cifar10 train ut (test_pipe. py). Under the same seed, the data of parameter and gradient updates is exactly equal to before this addition.

It can reduce  small memory on rank0  (stage0 contains embedding layer)

### CODE REVIEW

File: deepspeed/runtime/pipe/engine.py [hunk 1]
```diff
@@ -638,7 +638,6 @@ def _exec_forward_pass(self, buffer_id):
 
         # Zero out the gradients each time we use the tensor because only the data in
         # tensor changes across batches
-        self._zero_grads(inputs)
```
Comment (tohtana) [inline deepspeed/runtime/pipe/engine.py:641]: Can you explain why we can delete this? We could also delete the comment before this line.

File: deepspeed/runtime/pipe/module.py [hunk 2]
```diff
@@ -21,6 +21,13 @@
 from deepspeed.runtime.state_dict_factory import SDLoaderFactory
 from deepspeed.accelerator import get_accelerator
 from deepspeed.checkpoint.utils import clone_tensors_for_torch_save
+from enum import Enum
+
+
+class CkptLayer_Enum(Enum):
+    not_ckpt_layer = 0
+    normal_ckpt_layer = 1
+    warp_ckpt_layer = 2
```
Comment (tohtana) [inline deepspeed/runtime/pipe/module.py:30]: Does warp mean wrap?

File: deepspeed/runtime/pipe/module.py [hunk 3]
```diff
@@ -320,7 +343,7 @@ def forward(self, forward_input):
         # will see a different offset.
         self.micro_offset += 1
 
-        def exec_range_func(start, end):
+        def exec_range_func(start, end, warp_layer=False):
```
Comment (tohtana) [inline deepspeed/runtime/pipe/module.py:346]: Do you mean wrap, not warp?

File: deepspeed/runtime/pipe/module.py [hunk 4]
```diff
@@ -342,7 +365,29 @@ def exec_func(*inputs):
                     inputs = layer(inputs)
                 return inputs
 
-            return exec_func
+            def exec_func_warp(*inputs):
```
Comment (tohtana) [inline deepspeed/runtime/pipe/module.py:368]: Can we consolidate this with exec_func?
Most of the code are duplicated.

**Review by tohtana — COMMENTED** [review]
Thank you for submitting this PR, @inkcherry!
Is the goal of this to stop creating gradients of inputs when it is unnecessary?

Can you also clarify intentions of some changes?

**Comment by inkcherry** [comment]
> Thank you for submitting this PR, @inkcherry! Is the goal of this to stop creating gradients of inputs when it is unnecessary?
> 
> Can you also clarify intentions of some changes?

thanks for your review and apologies for not explaining clearly.

Yes, currently,  requires_grads of input of the first checkpoint layer must be true. I think this is a limitation, and the current code may have 2 following logic to work around this limitation in two scenarios.  please correct me if I'm wrong.
 
1 For image input, set ```requireds_ grad=true```  for input from the dataset, this will calculate and save grads for the input image, which I think may cost memory and computing time with large size input. related issue.
https://github.com/AGI-Collective/multimodal/issues/16
 
 
2 Secondly, for LLM, notice the comment https://github.com/microsoft/DeepSpeed/blob/64c670ef021118ed42a98d5dde249f35878a07a7/deepspeed/runtime/pipe/module.py#L621, it  hooks transformer layer(force make embedding layer as a non-checkpoint layer, 
which take input with requires_grad=False and ouput tensor with requires_grad=True). 
 
For this limitation, we could pass a dummy input that requires grad but isn't necessarily used in computation  FYI.
https://github.com/prigoyal/pytorch_memonger/blob/master/tutorial/Checkpointing_for_PyTorch_models.ipynb?short_path=e1e8ee7#L282C244-L282C257  , here wrapping a ``WrapModule``` at the first layer to achieve this.

File: deepspeed/runtime/pipe/module.py (hunk 2 above)
Comment (inkcherry) [reply deepspeed/runtime/pipe/module.py]: fixed.

File: deepspeed/runtime/pipe/module.py (hunk 3 above)
Comment (inkcherry) [reply deepspeed/runtime/pipe/module.py]: fixed.

File: deepspeed/runtime/pipe/engine.py (hunk 1 above)
Comment (inkcherry) [reply deepspeed/runtime/pipe/engine.py:641]: https://github.com/microsoft/DeepSpeed/blob/64c670ef021118ed42a98d5dde249f35878a07a7/deepspeed/runtime/pipe/engine.py#L620-L623  here inputs has become a non-leaf tensor, and its gradient has not been saved. if ```if self.is_pipe_partitioned and not self.is_first_stage()``` is False, not create a new leaf tensor named input,  access inputs' gradient in _zero_grad will trigger a warning.

File: deepspeed/runtime/pipe/module.py (hunk 4 above)
Comment (inkcherry) [reply deepspeed/runtime/pipe/module.py]: merged.

**Comment by inkcherry** [comment]
@tohtana , I've made some changes to the code. Any suggestions?  would appreciate your feedback.

**Comment by tohtana** [comment]
Hi @inkcherry, thank you for the fixes. Overall, this looks okay to me. 

On the other hand, I just reviewed #4118 and this might solve a part of the issue you addressed. I didn't notice this PR until very recently, and I'm sorry that I couldn't share it with you. 

We want to merge the PR, and I am wondering if we can simplify the changes in this PR using the new features proposed in #4118. Can you share your thoughts on this?

**Comment by inkcherry** [comment]
@tohtana, yes I agree, https://github.com/microsoft/DeepSpeed/pull/4118 is great！
I could try based on  https://github.com/microsoft/DeepSpeed/pull/4118 merged when it is merged.

**Comment by tjruwase** [comment]
@inkcherry, #4118 is now merged. Also, could you please add a unit test when you retry? Thanks!

**Comment by inkcherry** [comment]
moved to https://github.com/microsoft/DeepSpeed/pull/4224
close this one
